### DISCUSSION TITLE
Dependencies are installed multiple times for serverless functions

### DISCUSSION THREAD

**Discussion by erik-beus** [vercel/community#79] [Help] [answered]
We have a number of serverless functions, but when deploying to Vercel, it seems the dependencies (`npm install`) are installed once for every function?
That adds a lot of build time.

Is this how it's supposed to work? Or is there some way to avoid running `npm install` multiple times?

Thanks.

**Comment by styfle** [comment, marked as answer]
Indeed, this is expected. You could potentially have different dependencies for each function when using [`builds`](https://vercel.com/docs/configuration#project/builds) or with multiple `package.json` files.

The closest package.json to the serverless function is used, starting with the current directory and walking up the tree to the root of the project.

We might add an option in the future to disable that behavior to improve build performance when using multiple functions that all share the same dependencies.

But it remains for backwards compatibility.

**Reply by erik-beus** [reply]
I see.
I'm not sure what the most common project setup is, but if a lot of projects have the same dependencies for all functions, it might be worth detecting that and optimise the build performance automatically from Vercel.
We plan to have 20+ functions which will eventually lead to a pretty long build time (unless we try to only use the necessary dependencies for each of course). So any improvements on this would be appreciated :)

Anyway, thanks for the clarification 👍

**Reply by styfle** [reply]
I would suggest using Next.js which already has those optimizations when using the `/pages/api` directory.

https://nextjs.org/docs/api-routes/introduction

**Comment by robincwillis** [comment]
Is there any update on this? We're experiencing the same issues and incredible long build times. We're also already committed to Gatsby as a framework so we cannot switch to Next.js

**Comment by yachaka** [comment]
Solution would be to trigger specific builds for each function, depending on if a hash/source code has changed. I don't think Vercel provides this

I'm happy to have found this post, as I was searching for this in the documentation, didn't find it

**Comment by deve-sh** [comment]
Tried adding a `package.json` with only the dependencies required by the serverless Functions, seems to have reduced the build times but not sure if it's a one-off event since Vercel doesn't specify which package.json it's using. 

One thing is for sure, that Vercel doesn't try building package.json as a serverless function itself.

**Reply by Pieparker** [reply]
We appreciate you reaching out! Unfortunately, this topic is rather old and has been marked as Answered. If you’re exhibiting the same or a similar issue to the original poster, please feel free to create your own discussion and provide as much context as you can. This helps us to ensure that discussions are relevant, fresh and free of any possibly misleading context.
